lib/hexapdf/tokenizer.rb

```diff
@@ -439,9 +440,10 @@ def prepare_string_scanner(needed_bytes = nil)
       @io.seek(@next_read_pos)
       return false if @io.eof?
 
-      @ss << @io.read(8192)
-      if @ss.pos > 8192 && @ss.string.length > 16384
-        @ss.string.slice!(0, 8192)
+      @io.read(8192, @io_partial)
+      @ss << @io_partial
+      if @ss.pos > 8192 && @ss.string.bytesize > 16384
```

Comment:
Using `#bytesize` here should be equivalent to `#length` since the string is a binary string. However, it seems that `#length` is faster, so I would keep that:

~~~ yaml
prelude: |
  str = 'a'.b * 22345
benchmark:
  old: 'str.length'
  new: 'str.bytesize'
~~~

~~~
benchmark-driver bm1.yaml
Warming up --------------------------------------
                 old    47.258M i/s -     47.365M times in 1.002271s (21.16ns/i, 8clocks/i)
                 new    39.318M i/s -     39.532M times in 1.005455s (25.43ns/i, 10clocks/i)
Calculating -------------------------------------
                 old   195.591M i/s -    141.773M times in 0.724847s (5.11ns/i, 10clocks/i)
                 new    77.375M i/s -    117.953M times in 1.524427s (12.92ns/i, 26clocks/i)

Comparison:
                 old: 195590779.5 i/s
                 new:  77375439.1 i/s - 2.53x  slower
~~~